### ISSUE TITLE
Incorrect One-Hot Encoding Transformation in PMML Output Using sklearn2pmml

### ISSUE THREAD

**Issue by terryjunyang** [jpmml/sklearn2pmml#429]
Hello,

I've encountered an issue while using `sklearn2pmml` to convert a `OneHotEncoder` from `scikit-learn` into a PMML file. Instead of generating a one-hot encoded transformation, the PMML file contains an unexpected categorical regrouping transformation.

### Steps to Reproduce

1. **Dataset**: I used the [mpg dataset](https://github.com/ankane/eps/blob/fd2ddf2e8395d76395013117b7a7ec6c49c86edb/test/support/mpg.csv) with a categorical column `class`.
2. **Code**:
    
    ```python
    from sklearn.preprocessing import OneHotEncoder
    from sklearn2pmml import sklearn2pmml
    from pypmml import Model
    import pandas as pd
    import os
    
    save_dir = './'
    data = pd.read_csv("mpg.csv")
    data_1 = data[["class"]]
    print(data_1.head(10))
    print(data_1.value_counts())
    transformer = OneHotEncoder(sparse=False, max_categories=3)
    transformer.fit(data_1)
    
    sklearn2pmml(transformer, os.path.join(save_dir, "one_hot_tranformer_only.pmml"), with_repr=True)
    pmml_model = Model.load(os.path.join(save_dir, "one_hot_tranformer_only.pmml"))
    
    test_data = data_1.loc[[5]]
    input_data = test_data.to_dict(orient='records')[0]
    result_pmml = pmml_model.predict(input_data)
    
    print("Transformed data using transformer or pkl:")
    print(transformer.transform(test_data))
    print("\nTransformed data using PMML model:")
    print(result_pmml)
    ```
    
    Print output
    ```
            class
    0     midsize
    1  subcompact
    2         suv
    3         suv
    4     midsize
    5     compact
    6     compact
    7         suv
    8     compact
    9     compact
    class     
    suv           62
    compact       47
    midsize       41
    subcompact    35
    pickup        33
    minivan       11
    2seater        5
    dtype: int64
    Transformed data using transformer or pkl:
    [[1. 0. 0.]]
    
    Transformed data using PMML model:
    {'regroup(class)': 'compact'}
    ```
    
3. **Expected Behavior**:
The `class` column should be one-hot encoded in the PMML file, generating binary columns for each category.
4. **Actual Behavior**:
The PMML file contains a `DerivedField` that groups certain categories under an "infrequent" label instead of performing one-hot encoding. Below is the generated PMML:
  
    one_hot_tranformer_only.pmml
    ```xml
    <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
    <PMML xmlns="http://www.dmg.org/PMML-4_4" xmlns:data="http://jpmml.org/jpmml-model/InlineTable" version="4.4">
	    <Header>
		    <Application name="SkLearn2PMML package" version="0.110.0"/>
		    <Timestamp>2024-08-12T03:27:13Z</Timestamp>
	    </Header>
	    <DataDictionary>
		    <DataField name="class" optype="categorical" dataType="string">
			    <Value value="2seater"/>
			    <Value value="compact"/>
			    <Value value="midsize"/>
			    <Value value="minivan"/>
			    <Value value="pickup"/>
			    <Value value="subcompact"/>
			    <Value value="suv"/>
		    </DataField>
	    </DataDictionary>
	    <TransformationDictionary>
		    <DerivedField name="regroup(class)" optype="categorical" dataType="string">
			    <Apply function="if">
				    <Apply function="isIn">
					    <FieldRef field="class"/>
					    <Constant dataType="string">2seater</Constant>
					    <Constant dataType="string">midsize</Constant>
					    <Constant dataType="string">minivan</Constant>
					    <Constant dataType="string">pickup</Constant>
					    <Constant dataType="string">subcompact</Constant>
				    </Apply>
				    <Constant dataType="string">infrequent</Constant>
				    <FieldRef field="class"/>
			    </Apply>
		    </DerivedField>
	    </TransformationDictionary>
    </PMML>
    ```
    

### Environment

- **Python Version**: 3.11.5
- **scikit-learn Version**: 1.1.3
- **sklearn2pmml Version**: 0.110.0

### Additional Information

I expected the OneHotEncoder transformation to be directly translated into a one-hot encoded representation in the PMML file. Instead, the current behavior leads to a transformation that groups certain categories, which is not what was intended.

Could this be a bug, or is there a specific configuration or approach that I might be missing? Any guidance or fix would be greatly appreciated.

Thanks!
Terry.

**Comment by vruusmann** [comment]
> **Expected Behavior**: The class column should be one-hot encoded in the PMML file, generating binary columns for each category.

There is no one-hot encoding in PMML, because PMML models, transformations and functions can handle (categorical-) string columns as-is.

In contrast, Scikit-Learn models can't handle strings as strings, and requires them to be transformed into numeric representation using some encoding scheme.

> **Actual Behavior**: The PMML file contains a DerivedField that groups certain categories under an "infrequent" label.

The actual/observed behaviour is correct.

The "business logic" of a `OneHotEncoder(sparse=False, max_categories=3)` reads in human language: "keep the three most popular category levels as-is, replace all other (less popular-) category levels with a new "infrequent" pseudo-category".

From the PMML perspective, a one-hot encoding step in its default configuration is effectively a no-op. Its most visible "effect" is populating the `/PMML/DataDictionary/DataField` element with `Value@property="valid"` child elements, one for each category level.

In the current case, the one-hot necessitates generating a "regroup(class)" helper `DerivedField` element, because it has to replace some string category levels with the new "infrequent" category level.

But there is absolutely no need to replace string columns with binary vector columns.

**Comment by vruusmann** [comment]
Closing as "invalid" - the SkLearn2PMML package is doing correct job by leaving out no-op transformations (such as one-hot encoding).

> from pypmml import Model

BTW, consider using the [JPMML-Evaluator-Python](https://github.com/jpmml/jpmml-evaluator-python) library for evaluating PMML in Python. Your current tool is not good.

**Comment by fritshermans** [comment]
I noticed differences in predictions with pymml as well. The predictions made with an sklearn pipeline were not equal to prediction made with the pmml (created by sklearn2pmml) using the pymml package. When I used JPMML-Evaluator-Python, there were no differences
